cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -411,26 +415,231 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
       .collect(Collectors.toMap(JoinKey::getStageName, JoinKey::getFields));
+
+    Schema outputSchema = joinDefinition.getOutputSchema();
+    // if the output schema is null it means it could not be generated because some of the input schemas are null.
+    // there isn't a reliable way to get the schema from the data at this point, so we require the user to have
+    // provided the output schema directly
+    // it is technically possible to take a single StructuredRecord from each input to determine the schema,
+    // but this approach will not work if the input RDD is empty.
+    // in addition, RDD.take(1) and RDD.isEmpty() both trigger a Spark job, which is not desirable.
+    // when we properly propagate schema at runtime, this condition should no longer happen
+    if (joinDefinition.getOutputSchema() == null) {
+      throw new IllegalArgumentException(
+        String.format("Joiner stage '%s' cannot calculate its output schema because " +
+                        "one or more inputs have dynamic or unknown schema. " +
+                        "An output schema must be directly provided.", stageName));
+    }
     List<JoinCollection> toJoin = new ArrayList<>();
+    List<Schema> keySchema = null;
     while (stageIter.hasNext()) {
       // in this loop, information for each stage to be joined is gathered together into a JoinCollection
       JoinStage right = stageIter.next();
       String rightName = right.getStageName();
+      Schema rightSchema = right.getSchema();
+      List<String> key = stageKeys.get(rightName);
+      if (rightSchema == null) {
+        if (keySchema == null) {
+          keySchema = deriveKeySchema(stageName, stageKeys, joinDefinition);
+        }
+        // if the schema is not known, generate it from the provided output schema and the selected fields
+        rightSchema = deriveInputSchema(stageName, rightName, key, keySchema,
+                                        joinDefinition.getSelectedFields(), joinDefinition.getOutputSchema());
+      }
+
+      // drop fields that aren't included in the final output
+      Set<String> requiredFields = new HashSet<>(key);
+      for (JoinField joinField : joinDefinition.getSelectedFields()) {
+        if (!joinField.getStageName().equals(rightName)) {
+          continue;
+        }
+        requiredFields.add(joinField.getFieldName());
+      }
+      rightSchema = trimSchema(rightSchema, requiredFields);
+
       // JoinCollection contains the stage name,  SparkCollection, schema, joinkey,
       // whether it's required, and whether to broadcast
-      toJoin.add(new JoinCollection(rightName, inputDataCollections.get(rightName),
-                                    right.getSchema(), stageKeys.get(rightName),
+      toJoin.add(new JoinCollection(rightName, inputDataCollections.get(rightName), rightSchema, key,
                                     right.isRequired(), right.isBroadcast()));
     }
 
+    String leftName = left.getStageName();
```

Comment:
It is better to put this logic under https://github.com/cdapio/cdap/blob/fdf4d5c6fc991e44b35de436c17d067a72318a1f/cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java#L403, so that it is easier to know they are related